core/src/main/java/com/linecorp/armeria/server/Http1RequestDecoder.java

```diff
@@ -215,16 +245,28 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
                             final Routed<ServiceConfig> routed =
                                     routingCtx.virtualHost().findServiceConfig(routingCtx, true);
                             assert routed.isPresent();
+                            final ServiceConfig serviceConfig = routingCtx.result().value();
+                            if (serviceConfig.service().as(WebSocketService.class) != null &&
+                                isHttp1WebSocketUpgradeRequest(headers)) {
+                                logger.trace("Received WebSocket upgrade headers: {}", headers);
+                                this.req = req = DecodedHttpRequest.of(false, eventLoop, id, 1, headers,
+                                                                       keepAlive, inboundTrafficController,
+                                                                       routingCtx);
+                                // Because endOfStream is false and the exchangeType is BIDI_STREAMING.
+                                assert req instanceof StreamingDecodedHttpRequest;
+                                webSocketUpgradeContext.setLastWebSocketUpgradeRequestId(id);
```

Comment:
Question: Do we allow upgrading an HTTP/1.1 session to a WebSocket amid handling normal requests? I'm not 100% sure we can perfectly handle normal HTTP/1.1 requests and `WebSocket` frames in the same connection.

If a connection is shared and a response returns a `connection: close` header, the `WebSocket` connection might be closed as well which would be unexpected to users.

Reply:
> I'm not 100% sure we can perfectly handle normal HTTP/1.1 requests and WebSocket frames in the same connection.

I'm curious what kind of concerns you have. Before I took a look at this PR I was imagining websocket requests to be handled like a normal http request. (since it is technically a long-living http request)

I guess this is slightly related to how client-side is implemented. If we respond with a 400 here, I'm guessing we want to maintain separate connection pools for websockets since we don't want an intermittent 400 to be returned.

Reply:
One idea I was playing around with is whether we can just remove this decoder and add a handler dedicated for websockets. This way, we can modify `Http1RequestDecoder`  (and potentially `HttpResponseSubscriber`) without worrying about creating regressions for websocket. 

The downside of this approach is that some features (like `maxRequestLength`) won't be available out of the box and refactoring may be needed to support such features.

Another downside may be that if a websocket connection fails, the connection needs to be closed. (if a websocket upgrade request fails while reaching `WebSocketService`, the connection needs to be terminated)

I'm fine with the current approach (and I think we can switch later if needed) 😄    I just wanted to clarify what I was discussing earlier

https://github.com/line/armeria/commit/8667753b88b132b759c0f5b3c0d479194dbb0fae

Reply:
Updated to replace the ChannelHandler. Thanks!